src/index.js

```diff
@@ -1,135 +1,216 @@
+  return !value || Array.isArray(value) ? value : [value];
+}
+
+/** Backwards compatible version of `compiler.plugin.afterEmit.tapAsync()`. */
+function attachAfterEmitHook(compiler, callback) {
+  if (compiler.hooks) {
+    compiler.hooks.afterEmit.tapAsync('SentryCliPlugin', callback);
+  } else {
+    compiler.plugin('after-emit', callback);
   }
+}
 
-  if (originalEntry !== null && typeof originalEntry === 'object') {
-    const nextEntries = {};
-    Object.keys(originalEntry).forEach(key => {
-      nextEntries[key] = injectEntry(originalEntry[key], newEntry);
-    });
-    return nextEntries;
+/** Convenience function to add a webpack compilation error. */
+function addCompilationError(compilation, message) {
+  compilation.errors.push(`Sentry CLI Plugin: ${message}`);
+}
+
+class SentryCliPlugin {
+  constructor(options = {}) {
+    // By default we want that rewrite is true
+    this.options = Object.assign({ rewrite: true }, options);
+    this.options.include = toArray(options.include);
+    this.options.ignore = toArray(options.ignore);
+
+    this.cli = this.getSentryCli();
+    this.release = this.getReleasePromise();
   }
 
-  if (typeof originalEntry === 'string') {
-    return [newEntry, originalEntry];
+  /** Returns whether this plugin is in dryRun mode. */
+  isDryRun() {
+    return this.options.dryRun === true;
   }
 
-  if (typeof originalEntry === 'function') {
-    return () =>
-      Promise.resolve(originalEntry()).then(entry => injectEntry(entry, newEntry));
+  /** Creates a new Sentry CLI instance. */
+  getSentryCli() {
+    if (this.isDryRun()) {
+      return {
+        releases: {
+          proposeVersion: () => Promise.resolve('1.0.0-dev'),
+          new: () => Promise.resolve(),
+          uploadSourceMaps: () => Promise.resolve(),
+          finalize: () => Promise.resolve(),
+        },
+      };
+    }
+
+    return new SentryCli(this.options.configFile);
   }
 
-  return newEntry;
-}
+  /**
+   * Returns a Promise that will solve to the configured release.
+   *
+   * If no release is specified, it uses Sentry CLI to propose a version. The
+   * release string is always trimmed.
+   */
+  getReleasePromise() {
+    return (this.options.release
+      ? Promise.resolve(this.options.release)
+      : this.cli.releases.proposeVersion()
+    ).then(version => `${version}`.trim());
+  }
+
+  /** Checks if the given named entry point should be handled. */
+  checkEntry(key) {
+    const { entries } = this.options;
+    if (!entries) {
+      return true;
+    }
+
+    if (typeof entries === 'function') {
+      return entries(key);
+    }
 
-function injectRelease(compiler, versionPromise) {
-  if (typeof compiler.options === 'undefined') {
-    // Gatekeeper because we are not running in webpack env
-    // probably just tests
-    return;
+    if (entries instanceof RegExp) {
+      return entries.test(key);
+    }
+
+    if (Array.isArray(entries)) {
+      return entries.indexOf(key) >= 0;
```

Comment:
We can use `return entries.includes(key);` to be more descriptive

Reply:
Ah right, for some reason I wanted to be Node anno 1980 compatible -.-